big update

murych opened the pull request:
- рабочий `.gitignore`
- поправил зависимости
- скрипт для запуска
- инструкция для запуска через `cron`
- конфигурация приложения через `ini` файл

murych committed:
some architectural changes

- updated .gitignore to python pattern
- reading config vars with  configparser
- group and board ids are in config too
- updated README

murych committed:
some fixes

- fixed broken requirement
- updated README

murych committed:
updated README

murych committed:
run issues

- created bash script to execute things
- updated README

murych committed:
final updates

- fixed config file for correct execution from cron
- archiving posted cards
- little fixes to bash script

murych committed:
updated README

murych committed:
Merge branch 'master' of github.com:murych/postin-in-vk-from-trello

sgaynetdinov commented on main.py:
```diff
@@ -59,3 +59,4 @@ def is_card_can_published(trello_card):
         photo_items = Photo.upload_wall_photos_for_group(group.id, attachment_items.items())
         group.wall_post(message=card.name + '\n' + card.description, attachments=photo_items)
         card.set_due_complete()
+        card.set_closed(True)
```
Для чего нужно?

murych replied:
Карточки, просто отмеченные как "выполненные" постятся заново при запуске скрипта. Архивация для избежания подобного поведения.

sgaynetdinov replied:
> Карточки, просто отмеченные как "выполненные" постятся заново при запуске скрипта

нужно логику проверит, то самостоятельно отправлять в "архив" не нужно

sgaynetdinov replied:
> постятся заново при запуске скрипта

где то используется ?

sgaynetdinov commented on README.md:
````diff
@@ -1,4 +1,62 @@
+sudo pip3 install virtualenv virtualenvwrapper
+```
+
+- Point it to always run `python3` executable
````
Почему только python3?

sgaynetdinov commented on config.py:
```diff
@@ -1,3 +1,19 @@
+import configparser
+
+config = configparser.ConfigParser()
+config_file = os.path.join(os.path.dirname(__file__), 'config.ini')
```
Зачем настройки для python, хранить в `ini` ?
когда можно просто импортировать в проект `import config`

murych replied:
- легче читается 
- не нужно прописывать `.py` файлы в `.gitignore`

sgaynetdinov replied:
ты создаешь `config.ini`, который заполняет `config.py`, когда можно всё хранить в `config.py`

> не нужно прописывать .py файлы в .gitignore

можно прописать `config.py`

murych replied:
Некоторое время назад, мой проект на Django не хотел импортировать настройки из `.py` файла, у меня своеобразная аллергия на этот метод. `ini` файлы выступают хорошей альтернативой наряду с переменными окружения.
В принципе, это ведь можно и изменить назад. Главной идеей было вынесение "захардкоденных" параметров в конфиг.

sgaynetdinov replied:
> Некоторое время назад, мой проект на Django не хотел импортировать настройки из .py файла

а в чём дело было?

> Главной идеей было вынесение "захардкоденных" параметров в конфиг.

Идея хорошая 👍 , но лучше хранить настройки приложения python в файле `.py`, так поступают многие проекты.

murych replied:
> так поступают многие проекты

Что, к сожалению, не всегда говорит о коррекности таких действий. Не буду строить из себя опытного разработчика, но, например, [официальная документация к Django](https://docs.djangoproject.com/en/2.0/howto/deployment/checklist/#secret-key) вообще рекомендует использовать переменные окружения или _текстовые файлы_. 

Просто хранить API ключ от контакта хочется более-менее аккуратно. `ini` файлы, на мой взгляд, предлагают довольно строгий и аккуратный способ. [Некоторые](https://stackoverflow.com/a/23446849/2079040) предлагают использовать `yaml`, скорее всего, это уже дело вкуса.

> а в чём дело было?

Хотел бы я знать, честное слово. Так и не разобрались спустя почти год.

sgaynetdinov replied:
> Что, к сожалению, не всегда говорит о коррекности таких действий. Не буду строить из себя опытного разработчика, но, например, [официальная документация к Django] вообще рекомендует использовать переменные окружения или текстовые файлы.

можно ссылку

murych replied:
Упс, недоглядел. Исправил комментарий, но ссылку продублирую тут -> https://docs.djangoproject.com/en/2.0/howto/deployment/checklist/#secret-key

sgaynetdinov replied:
Тебя не настораживает, что они тут пишут только про одну переменную `SECRET_KEY`, а про остальные не слова. Думаю ты неверно понял посыл, тут не говориться, что  против хранения переменный в `settings.py`, они предостерегают его раскрытия 3-ем лицам, т.к. `SECRET_KEY` используется в качестве "соли".

murych replied:
В руководстве по VK API при выдаче секретного ключа написано, что при определенных обстоятельствах, аккаунт может быть заблокирован. Из-за чего я бы приравнял уровень безопасности хранения API ключей к `SECRET_KEY` в django. Остальные же параметры можно хранить в конфиге для удобства.

В принципе я понимаю, почему настройки часто хранят в `py` файлах.

sgaynetdinov replied:
> В руководстве по VK API при выдаче секретного ключа написано, что при определенных обстоятельствах, аккаунт может быть заблокирован.

vk.com не важно как и где хранишь свой "ключ", эти строчки не о хранение ключа, о скорее о том в каких целях ты будешь им пользоваться

> Из-за чего я бы приравнял уровень безопасности хранения API ключей к SECRET_KEY в django. 

django за тебя не хранит "ключ" где-то в файле, также и тут, разработчик который будет пользоваться этим скриптом/программой сам позаботиться где будет его "ключ", за него лучше ни чего не решать.

sgaynetdinov commented on run.sh:
```diff
@@ -0,0 +1,6 @@
+RUN_SCRIPT=/path/to/project/main.py
+source $ACTIVATE_SCRIPT
+python $RUN_SCRIPT
+echo "done"
```
зачем этот файл?
можно запустить `python main.py`

sgaynetdinov commented on README.md:
````diff
@@ -1,4 +1,62 @@
+- Point it to always run `python3` executable
+```bash
+echo "export VIRTUALENVWRAPPER_PYTHON=/usr/bin/python3" >> ~/.bashrc
+echo "source /usr/local/bin/virtualenvwrapper.sh" >> ~/.bashrc
````
Показывать как установить `virtualenv` и `virtualenvwrapper` и настройки не нужно

sgaynetdinov commented on README.md:
```diff
@@ -1,4 +1,62 @@
+
+[VK]
+ACCESS_TOKEN = AAAAAAAA
+GROUP = AAAAAAAA
```
настройки лучше хранить в `config.py`

sgaynetdinov commented on README.md:
````diff
@@ -1,4 +1,62 @@
+```bash
+workon trello-vk  # in case you've logged out
+python /path/to/your/project/main.py
+```
````
Думаю эти строчки избыточны?

murych replied:
Хотел написать инструкцию максимально доступную человеку, впервые взявшему в руки Python. Эта же причина для остальных избыточных пояснений.

sgaynetdinov replied:
Ему лучше смотреть в другом месте, всё таки `README` нужна что показать как запустить данный скрипт/программу/приложение, а не на учится разным фишкам.

sgaynetdinov replied:
в python3 можно не устанавливать `virtualenv`, она уже внутри `python -m venv my_venv`

murych replied:
Согласен, эта часть действительно избыточна

sgaynetdinov commented on README.md:
````diff
@@ -1,4 +1,62 @@
+and add this:
+```
+ */5 * * * * bash /path/to/repo/run.sh
+```
````
Может лучше написать "вечный" цикл на python, с таймером, для примера.
Кому нужно и так будут использовать свой `schedule`

murych replied:
Не проводил измерений, но подозреваю, что разово запускаемый скрипт ест меньше оперативной памяти, чем постоянно запущенный интерпретатор.

sgaynetdinov replied:
Дело не в памяти, кто захочет запустить, тот сам знает как переодически запускать такие задачи.

У меня данный скрипт на heroku работает, там свой cron.

murych replied:
Я для heroku нашел только плагин Temporize, который по cron-у умеет только GET и POST запросы отправлять

sgaynetdinov replied:
https://devcenter.heroku.com/articles/clock-processes-python

murych replied:
Спасибо за наводку. На мой взгляд `cron` задача куда более прозрачна. Да и не получится на бесплатном heroku аккаунте в рамках одного приложения запустить одновременно `worker` и `clock`

sgaynetdinov replied:
что такое `worker` и `clock` ?
и зачем их запускать одновременно?

murych replied:
Извиняюсь, более вдумчивое чтение документации показывает, что [можно](https://devcenter.heroku.com/articles/scheduled-jobs-custom-clock-processes#defining-custom-clock-processes)

`worker` и `clock` (а еще `web`) -- это [названия для процессов](https://devcenter.heroku.com/articles/how-heroku-works#knowing-what-to-execute), запускаемых в heroku приложении. Обычно `worker` и `web` используют для запуска самого приложения (например, какого-нибудь бота). Документация предлагает использовать название `clock` для обозначения процесса, выполняемого по расписанию.

murych replied:
Одновременно запускать для того, чтобы в рамках одного приложения (блога, например) некоторые функции запускать независимо от главного процесса.

sgaynetdinov replied:
Лучше бота назвать `bot`, а не `web` )))

sgaynetdinov replied:
для этого скрипта нужен только один "процесс", а как ты его назовешь это не важно

sgaynetdinov commented on README.md:
```diff
@@ -1,4 +1,62 @@
 ![N|Solid](https://img.shields.io/pypi/l/py-vkontakte.svg)
-![N|Solid](https://img.shields.io/pypi/pyversions/py-vkontakte.svg)
+[![Codacy Badge](https://api.codacy.com/project/badge/Grade/21c92d804ab541c8ae219b7098df3173)](https://www.codacy.com/app/murych/postin-in-vk-from-trello?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=murych/postin-in-vk-from-trello&amp;utm_campaign=Badge_Grade)
```
А зачем удалил `![N|Solid](https://img.shields.io/pypi/pyversions/py-vkontakte.svg)` ?

murych replied:
Он же показывает совместимость библиотеки с разными версиями Python, а не скрипта.

sgaynetdinov replied:
точно, не доглядел

murych committed:
pull request fixes

1. removed useless readme section and run script
2. updated "is_card_can_published" function logic so that posted cards can't be post again
3. config.py is the only way to configure app

murych commented:
Убрал лишние инструкции и немного дополнил механизм работы скипта, чтобы он не отправлял в вк карточки с отмеченным `due` :white_check_mark:
